File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums)<=1:return 0
        totaljumps=0
        lastindex=len(nums)-1
        jumpfrom=lastindex-1
        while jumpfrom>=0:
            if jumpfrom+nums[jumpfrom]>=lastindex :
                i=0
                while i<jumpfrom:
                    if i+nums[i]>=lastindex and i<jumpfrom:
                        jumpfrom=i
                        break
                    i+=1
                lastindex=jumpfrom
                totaljumps+=1
            jumpfrom-=1
        return totaljumps

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n=nums.size();
        vector<int> dp(n,INT_MAX);
        dp[n-1]=0;
        // calc(nums,dp,n-2);
        for (int i=n-2;i>=0;i--){
            int min=dp[i];
            for (int j=i+1;j<=i+nums[i] && j<n;j++){
                // if (j>=n){
                //     break;
                // }
                if (dp[j]==INT_MAX){
                    continue;
                }
                dp[i]=std::min(dp[i],1+dp[j]);
            }
        }
        return dp[0];
    }
};